web-crawler/src/main/java/com/google/sps/webcrawler/WebCrawler.java

```diff
@@ -0,0 +1,231 @@
+    }
+  }
+
+  /** 
+   * Checks robots.txt for permission to web-scrape, scrapes webpage if permitted and extracts
+   * textual content.
+   */
+  public NewsArticle scrapeAndExtractHtml(String url) {
+    String robotsUrl = url.substring(0, url.indexOf("/", url.indexOf("//") + 2) + 1)
+        + "robots.txt";
+    try {
+      InputStream robotsTxtStream = new URL(robotsUrl).openStream();
+      RobotsTxt robotsTxt = RobotsTxt.read(robotsTxtStream);
+      String webpagePath = url.substring(url.indexOf("/", url.indexOf("//") + 2));
+      Grant grant = robotsTxt.ask("*", webpagePath);
+      // Check and set crawl delay.
+      if (grant != null && grant.getCrawlDelay() != null) {
+        if (nextAccessTimes.containsKey(robotsUrl)) {
+          // Wait until the crawl delay fully passes.
+          if (System.currentTimeMillis() < nextAccessTimes.get(robotsUrl)) {
+            if (!waitForCrawlDelay(nextAccessTimes.get(robotsUrl) - System.currentTimeMillis())) {
+              return new NewsArticle();
+            };
+          }
+          nextAccessTimes.replace(robotsUrl,
+                                  System.currentTimeMillis() + grant.getCrawlDelay() * 1000);
+        } else {
+          System.out.println(System.currentTimeMillis());
+          nextAccessTimes.put(robotsUrl,
+                              System.currentTimeMillis() + grant.getCrawlDelay() * 1000);
+        }
+      }
+      // Check permission to access.
+      if (grant == null || grant.hasAccess()) {
+        InputStream webpageStream = new URL(url).openStream();
+        return ContentExtractor.extractContentFromHtml(webpageStream, url);
+      } else {
+        return new NewsArticle();
+      }
+    } catch (IOException e) {
+      System.out.println("[ERROR] Error occured during web scraping.");
+      return new NewsArticle();
+    }
+  }
+
+  /** 
+   * Waits for {@code timeToDelay} milliseconds and returns true if the pause was successful.
+   */
+  private boolean waitForCrawlDelay(long timeToDelay) {
+    try {
+      TimeUnit.MILLISECONDS.sleep(timeToDelay);
+    } catch (InterruptedException e) {
+      return false;
+    }
+    return true;
+  }
+
+  // @TODO [Fill in other properties: published date, publisher.]
+  /** 
+   * Stores {@code NewsArticle}'s metadata and content into the database, following a predesigned
+   * database schema.
+   * Requires "gcloud config set project project-ID" to be set correctly. Assumes "content" to be
+   * occupy fewer than 1500 bytes.
+   */
+  public void storeInDatabase(String candidateId, NewsArticle newsArticle) {
+    Datastore datastore = DatastoreOptions.getDefaultInstance().getService();
+    Key newsArticleKey = datastore.newKeyFactory()
+        .setKind("NewsArticle")
+        .newKey((new Random()).nextLong());
+    Entity newsArticleEntity = Entity.newBuilder(newsArticleKey)
+        .set("candidateId", datastore.newKeyFactory().setKind("Candidate").newKey(candidateId))
```

Comment:
From my observation, I think news articles are mostly about one particular candidate. If an article mentions another candidate, it usually only does so briefly. I imagine that the RelevancyChecker would give different salience scores as well.

Reply:
However, I believe it's definitely necessary to avoid duplication in the database. Right now each news article has a single foreign key pointing to one candidate. I can adjust the data structure to an array of candidate IDs

Reply:
If this would be rare, then no need to alter your entire database schema to accommodate.  It's probably fine if there are only a few duplicates for certain candidates :)

Reply:
Got it! But I'll make sure to keep this in mind if the situation changes